[Metricbeat][vSphere] Support for configurable IntervalId for performance API (#40678)
elastic/beats: metricbeat/docs/modules/vsphere.asciidoc

[[metricbeat-module-vsphere]]
== vSphere module

> The vSphere module uses the https://github.com/vmware/govmomi[Govmomi] library to collect metrics from any VMware SDK URL (ESXi/VCenter). This library is built for and tested against ESXi and vCenter 5.5, 6.0, and 6.5.

Reviewer: Broken link

> The vSphere module uses the https://github.com/vmware/govmomi[Govmomi] library to collect metrics from any VMware SDK URL (ESXi/VCenter). This library is built for and tested against ESXi and vCenter 5.5, 6.0, and 6.5.

Author: updated

> The vSphere module uses the https://github.com/vmware/govmomi[Govmomi] library to collect metrics from any VMware SDK URL (ESXi/VCenter). This library is built for and tested against ESXi and vCenter 5.5, 6.0, and 6.5.

Reviewer: We have now tested our code with the customer vsphere Client which is version 7.0.3. 
I think we can mention that testing has been done up to 7.0.3 version. 
@kush-elastic

> The vSphere module uses the https://github.com/vmware/govmomi[Govmomi] library to collect metrics from any VMware SDK URL (ESXi/VCenter). This library is built for and tested against ESXi and vCenter 5.5, 6.0, and 6.5.

Author: Sure, got it.

> By default, the vSphere module enables the following metric sets:

Reviewer 2: Please do `metricsets` everywhere!
```suggestion
By default, the vSphere module enables the following metricsets:
```

1. cluster

2. datastore

3. datastorecluster

4. host

5. network

6. resourcepool

7. virtualmachine

[float]
=== Supported Periods:

> The Datastore and Host metricsets support performance data collection using the vSphere performance API. Since the performance API has usage restrictions based on data collection intervals, users should ensure that the period is configured optimally to receive real-time data. This configuration can be determined based on the https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.monitoring.doc/GUID-247646EA-A04B-411A-8DD4-62A3DCFCF49B.html[Data Collection Intervals] and https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.monitoring.doc/GUID-25800DE4-68E5-41CC-82D9-8811E27924BC.html[Data Collection Levels].

Reviewer: ```suggestion
The Datastore and Host metricsets support performance data collection using the vSphere performance API. Given that the performance API imposes usage restrictions based on data collection intervals, users should configure the period optimally to ensure the receipt of real-time data. This configuration can be determined based on the https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.monitoring.doc/GUID-247646EA-A04B-411A-8DD4-62A3DCFCF49B.html[Data Collection Intervals] and https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.monitoring.doc/GUID-25800DE4-68E5-41CC-82D9-8811E27924BC.html[Data Collection Levels].
```

[IMPORTANT]

Only host and datastore metricsets have limitation of period restrictions.

[float]
==== Real-time data collection interval:
- 20

[float]
==== Historical data collection interval:

> - 300

Reviewer: ```suggestion
- 300s
```

Also add seconds after all time periods

[[metricbeat-module-vsphere]]
== vSphere module

The vSphere module uses the https://github.com/vmware/govmomi[Govmomi] library to collect metrics from any VMware SDK URL (ESXi/VCenter).

This module has been tested against ESXi and vCenter versions 5.5, 6.0, 6.5, and 7.0.3.

By default, the vSphere module enables the following metricsets:

1. cluster

2. datastore

3. datastorecluster

4. host

5. network

6. resourcepool

7. virtualmachine

[float]
=== Supported Periods:
The Datastore and Host metricsets support performance data collection using the vSphere performance API. Given that the performance API imposes usage restrictions based on data collection intervals, users should configure the period optimally to ensure the receipt of real-time data. This configuration can be determined based on the https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.monitoring.doc/GUID-247646EA-A04B-411A-8DD4-62A3DCFCF49B.html[Data Collection Intervals] and https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.monitoring.doc/GUID-25800DE4-68E5-41CC-82D9-8811E27924BC.html[Data Collection Levels].

[IMPORTANT]

> Only host and datastore metricsets have limitation of period restrictions.

Author: I think later one is better.

- module: vsphere
  enabled: true
  metricsets: ["cluster", "datastore", "datastorecluster", "host", "network", "resourcepool", "virtualmachine"]
  
  # Real-time data collection – An ESXi Server collects data for each performance counter every 20 seconds by default.
  # Supported Periods:

>   # The Datastore and Host metricsets support performance data collection using the vSphere performance API.

Reviewer: shall we also mention that this will not impact the metrics collection other than perf metrics.

Author: Nice idea. let me do that.
